File: nodux_account_voucher/account_voucher/payment_entry_custom.py
# -*- coding: utf-8 -*-
# Copyright (c) 2015, Frappe Technologies Pvt. Ltd. and contributors
# For license information, please see license.txt

from __future__ import unicode_literals
import frappe, json
from frappe import _, scrub, ValidationError
from frappe.utils import flt, comma_or, nowdate
from erpnext.accounts.utils import get_outstanding_invoices, get_account_currency, get_balance_on
from erpnext.accounts.party import get_party_account
from erpnext.accounts.doctype.journal_entry.journal_entry \
	import get_average_exchange_rate, get_default_bank_cash_account
from erpnext.setup.utils import get_exchange_rate
from erpnext.accounts.general_ledger import make_gl_entries

from erpnext.controllers.accounts_controller import AccountsController

def validate(doc, event):
    setup_party_account_field(doc)
    set_missing_values(doc)
    validate_payment_type(doc)
    validate_party_details(doc)
    validate_bank_accounts(doc)
    set_exchange_rate(doc)
    validate_mandatory(doc)
    validate_reference_documents(doc)
    set_amounts(doc)
    clear_unallocated_reference_document_rows(doc)
    validate_payment_against_negative_invoice(doc)
    validate_transaction_reference(doc)
    set_title(doc)
    set_remarks(doc)


def validate_bank_accounts(doc):
	if doc.payment_type in ("Pay", "Internal Transfer"):
		for payment_form in doc.payments_type:
			validate_account_type(doc.mode_of_payment, ["Bank", "Cash"])

    if doc.payment_type in ("Receive", "Internal Transfer"):
    	for payment_form in doc.payments_type:
            validate_account_type(doc.paid_to, ["Bank", "Cash"])

def validate_account_type(account, account_types):
	account_type = frappe.db.get_value("Account", account, "account_type")
	if account_type not in account_types:
		frappe.throw(_("Account Type for {0} must be {1}").format(account, comma_or(account_types)))

File: nodux_account_voucher/account_voucher/payment_entry_custom.js
// Copyright (c) 2016, Frappe Technologies Pvt. Ltd. and contributors
// For license information, please see license.txt
{% include "erpnext/public/js/controllers/accounts.js" %}

frappe.ui.form.on('Payment Entry Type', {

	payment_form: function(frm, cdt, cdn) {
		var d = locals[cdt][cdn];
		get_payment_mode_account(frm, d.payment_form, function(account){
			frappe.model.set_value(cdt, cdn, 'account', account)
		});
		if(d.payment_form == "Cheque"){
			frappe.model.set_value(cdt, cdn, 'titular', frm.doc.party);
		}
	},

	amount: function(frm) {
		var total_amount = 0.0;
		$.each(frm.doc.payments_type || [], function(i, row) {
				if (row.amount) {
				total_amount += flt(row.amount);
			}
		});
		frm.set_value("paid_amount", total_amount);
	},

	account: function(frm){
		var mound, mode;
		$.each(frm.doc.payments_type || [], function(i, row) {
				if (row.account) {
					mound = row.account;
					mode = row.payment_form;
			}
		});
		if(frm.doc.payment_type == "Receive") {
			frm.set_value("paid_to", mound);
		}else if (frm.doc.payment_type == "Pay") {
			frm.set_value("paid_from", mound);
		}
	},
 // CUANDO SE AGREGUE DOCTYPE BANCO
	// bank:function(frm, cdt, cdn){
	// 	var d = locals[cdt][cdn];
	// 	var nro_account = frappe.db.get_value("Bank", {"bank_name": d.bank.}, "nro_account");
	// 	var nro_document = frappe.db.get_value("Bank", {"bank_name": d.bank}, "nro_document");
	//	frappe.model.set_value(cdt, cdn, 'nro_account', nro_account);
	//	frappe.model.set_value(cdt, cdn, 'nro_document', nro_document);
	// },

	nro_document:function(frm){
		var num_ref;
		$.each(frm.doc.payments_type || [], function(i, row) {
				if (row.nro_document) {
					num_ref = row.nro_document;
			}
		});
		frm.set_value("reference_no", num_ref);
	},

	date_reference:function(frm){
		var date_ref;
		$.each(frm.doc.payments_type || [], function(i, row) {
				if (row.date_reference) {
					date_ref = row.date_reference;
			}
		});
		frm.set_value("reference_date", date_ref);
	}
})
